executor/pipeline/insert.rs

```diff
@@ -110,20 +111,17 @@ where
 
 pub(crate) fn prepare_output_rows(
     output_width: u32,
-    input_iterator: impl StageIterator,
+    mut input_iterator: impl StageIterator,
     mapping: &Vec<(VariablePosition, VariablePosition)>,
 ) -> Result<Batch, Box<PipelineExecutionError>> {
-    // TODO: if the previous stage is not already in Collected format, this will end up temporarily allocating 2x
-    //       the current memory. However, in the other case we don't know how many rows in the output batch to allocate ahead of time
-    //       and require resizing. For now we take the simpler strategy that doesn't require resizing.
-    let input_batch = input_iterator.collect_owned()?;
-    let total_output_rows = input_batch.get_multiplicities().iter().sum::<u64>() as usize;
-    let mut output_batch = Batch::new(output_width, total_output_rows);
-    let mut input_batch_iterator = input_batch.into_iterator();
-    while let Some(row) = input_batch_iterator.next() {
+    let initial_output_batch_size = input_iterator
+        .multiplicity_sum_if_collected()
+        .map(|sum| 10 + sum + (sum / 8)) // Some breathing room to avoid re-allocation
+        .unwrap_or(BATCH_DEFAULT_CAPACITY);
+    let mut output_batch = Batch::new(output_width, initial_output_batch_size);
+    while let Some(row) = input_iterator.next().transpose()? {
```

Comment:
The `collect_owned()` on (the deleted) line 119 would do the resizing we were afraid of anyway.
We can avoid the resizing cost when the previous stage is already collected, by checking and finding the multiplicity sum .